src/lfortran/semantics/ast_common_visitor.h

```diff
@@ -1187,6 +1187,129 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
+
+        current_scope = current_scope_copy;
+        current_scope->add_symbol(func_name, func_sym);
+
+        ASR::expr_t* func_call = ASRUtils::EXPR(ASRUtils::make_FunctionCall_t_util(al, end_sym->base.loc,
+                                func_sym, func_sym, nullptr, 0, ASRUtils::symbol_type(end_sym), nullptr, nullptr, false));
+        return func_call;
+    }
+
+    ASR::expr_t* convert_integer_binop_to_function_call(ASR::expr_t* end, bool is_argument) {
+        ASR::IntegerBinOp_t* end_bin_op = ASR::down_cast<ASR::IntegerBinOp_t>(end);
+        ASR::expr_t* left = end_bin_op->m_left;
+        ASR::expr_t* right = end_bin_op->m_right;
+        if (ASR::is_a<ASR::Var_t>(*left) && !ASR::is_a<ASR::Var_t>(*right)) {
+            /* 
+            Handle expressions like `nx + a` where `a` can either be 
+            an integer or an `IntegerBinOp_t`.
+
+            Examples - nx + 1 and nx + ny * nz
+            */
+            ASR::Var_t* end_var = ASR::down_cast<ASR::Var_t>(left);
+            ASR::symbol_t* end_sym = end_var->m_v;
+            SymbolTable* symbol_scope = ASRUtils::symbol_parent_symtab(end_sym);
+            if (ASR::is_a<ASR::ExternalSymbol_t>(*end_sym) ||
+                (symbol_scope->counter != current_scope->counter && is_argument &&
+                ASRUtils::expr_value(end) == nullptr) ) {
+                left = get_transformed_function_call(end_sym);
+            }
+            if (ASR::is_a<ASR::IntegerBinOp_t>(*right)) {
+                right = convert_integer_binop_to_function_call(right, is_argument);
+            }
+        } else if (!ASR::is_a<ASR::Var_t>(*left) && ASR::is_a<ASR::Var_t>(*right)) {
+            /* 
+            Handle expressions like `a + nx` where `a` can either be 
+            an integer or an `IntegerBinOp_t`.
+
+            Examples - 1 + nx and ny * nz + nx
+            */
+            ASR::Var_t* end_var = ASR::down_cast<ASR::Var_t>(right);
+            ASR::symbol_t* end_sym = end_var->m_v;
+            SymbolTable* symbol_scope = ASRUtils::symbol_parent_symtab(end_sym);
+            if (ASR::is_a<ASR::ExternalSymbol_t>(*end_sym) ||
+                (symbol_scope->counter != current_scope->counter && is_argument &&
+                ASRUtils::expr_value(end) == nullptr) ) {
+                right = get_transformed_function_call(end_sym);
+            }
+            if (ASR::is_a<ASR::IntegerBinOp_t>(*left)) {
+                left = convert_integer_binop_to_function_call(left, is_argument);
+            }
+        } else if (ASR::is_a<ASR::Var_t>(*left) && ASR::is_a<ASR::Var_t>(*right)) {
+            // Handle expressions like `nx + ny` where both `nx` and `ny` are 
+            // external variables.
+            ASR::symbol_t* first_end_sym = (ASR::symbol_t*)ASR::down_cast<ASR::Var_t>(left)->m_v;
+            ASR::symbol_t* second_end_sym = (ASR::symbol_t*)ASR::down_cast<ASR::Var_t>(right)->m_v;
```

Comment:
```suggestion
            ASR::symbol_t* first_end_sym = ASR::down_cast<ASR::Var_t>(left)->m_v;
            ASR::symbol_t* second_end_sym = ASR::down_cast<ASR::Var_t>(right)->m_v;
```
I think both are already `symbol_t*`. Unsure if you need to explicitly cast them.

Reply:
You are right. :+1: